pkg/efivarfs/varfs.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2014-2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// This code is a somewhat simplified and changed up version of
+// github.com/canonical/go-efilib credits go their authors.
+// It allows interaction with the efivarfs via u-root which means
+// both read and write support.
+
+package efivarfs
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"syscall"
+
+	"github.com/u-root/u-root/pkg/guid"
+	"golang.org/x/sys/unix"
+)
+
+func init() {
+	if !probeEfivarfs() {
+		return
+	}
+}
+
+// EfiVarFs is the path to the efivarfs mount point
+const EfiVarFs = "/sys/firmware/efi/efivars/"
+
+var vars varsBackend = varfsBackend{}
+
+type varsBackend interface {
+	Get(name string, guid guid.GUID) (VariableAttributes, []byte, error)
+	Set(name string, guid guid.GUID, attrs VariableAttributes, data []byte) (bool, error)
+	Delete(name string, guid guid.GUID) error
+	List() ([]VariableDescriptor, error)
+}
+
+type varfsBackend struct{}
+
+// Get reads the contents of an efivar if it exists and has the necessary permission
+func (v varfsBackend) Get(name string, guid guid.GUID) (VariableAttributes, []byte, error) {
+	path := filepath.Join(EfiVarFs, fmt.Sprintf("%s-%s", name, guid))
+	f, err := openVarfsFile(path, os.O_RDONLY, 0)
+	switch {
+	case os.IsNotExist(err):
+		return 0, nil, ErrVarNotExist
+	case os.IsPermission(err):
+		return 0, nil, ErrVarPermission
+	case err != nil:
+		return 0, nil, err
+	}
+	defer f.Close()
+
+	var attrs VariableAttributes
+	err = binary.Read(f, binary.LittleEndian, &attrs)
+	if err != nil {
+		if err == io.EOF {
+			return 0, nil, ErrVarNotExist
+		}
+		return 0, nil, err
+	}
+
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return 0, nil, err
+	}
+	return attrs, data, nil
+}
+
+// Set modifies a given efivar with the provided contents
+func (v varfsBackend) Set(name string, guid guid.GUID, attrs VariableAttributes, data []byte) (bool, error) {
```

Comment:
What is the purpose of the bool? It seems to be always false?

Reply:
There is one possibility for it to be true in the code, which is described in the code comments a few lines down, in short: It triggers a retry.